File: nexus/ui/src/api/audit/forget.ts
import { fetchJson, readCsrfHeader } from "@nube/starter-client-ts";
import type { StarterClient } from "@nube/starter-client-ts";

import type { ForgetRequest, ForgetResponse } from "@/api/types";

// `POST /api/v1/audit/forget` — GDPR right-to-erasure for a user subject
// (admin-gated). Tombstones the `before`/`after`/`patch` of every change the
// subject authored in the caller's tenant while keeping the audit fact (who,
// when, which op). Returns the number of rows scrubbed.
export function forgetSubject(
  client: StarterClient,
  request: ForgetRequest,
): Promise<ForgetResponse> {
  return fetchJson<ForgetResponse>(client, `${client.apiPrefix}/audit/forget`, {
    method: "POST",
    headers: { "content-type": "application/json", ...readCsrfHeader() },
    body: JSON.stringify(request),
  });
}

File: nexus/backend/crates/nexus-spi/src/dto/audit/forget.rs
//! Request/response of `POST /api/v1/audit/forget` — GDPR right-to-erasure.

use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

/// Which user to erase from the audit ledger. Erasure tombstones the *content*
/// (`before`/`after`/`patch`) of every change the subject authored, within the
/// caller's tenant; the audit fact (who/when/what op) is preserved, as a
/// regulator still needs to see that an action occurred.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ForgetRequest {
    /// The principal subject (stable user id) whose authored changes to scrub.
    pub subject: String,
}

/// Result of a forget request: how many ledger rows were tombstoned.
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ForgetResponse {
    /// Number of ledger rows whose payloads were nulled.
    pub tombstoned: u64,
}
